criu/kerndat.c

```diff
@@ -917,9 +917,9 @@ static int kerndat_uffd(void)
 	 * have a problem!"
 	 */
 	if (uffd < 0) {
-		if (uffd == -ENOSYS)
+		if (uffd == -ENOSYS-1)
```

Comment:
I'd prefer `"int *err"` argument to `uffd_open`.
Now looking at it, I'm not sure we need to fail at all if `uffd_open()` returns error. Apparently, simply stetting `kdat.has_uffd` to false if `uffd_open()` failed would be sufficient.
Probably we'd also need to make sure that `opts.lazy_pages` and `kdat.has_uffd` match, somewhere in `main()`...

Reply:
Me too. @Snorch , please

Reply:
`"int *err"` part done.

Not sure if you instead mean to

>simply stetting kdat.has_uffd to false if uffd_open() failed would be sufficient

Please correct me if I did wrong one from two options =)

Reply:
> Not sure if you instead mean to
> 
> > simply stetting kdat.has_uffd to false if uffd_open() failed would be sufficient
> 
> Please correct me if I did wrong one from two options =)

I was mostly thinking aloud :)
The point is that we don't really care if we can use userfaultfd unless user explicitly asked for lazy pages. So it does not matter what was the exact reason `uffd_open()` failed and we can always return 0 from `kerndat_uffd()`. The availability of userfault will be set in `kdat.has_uffd`. If the user didn't ask for lazy-pages we can continue perfectly fine even if userfault is not available and only fail restore if both `kdat.has_uffd = false` and `opts.lazy_pages = true`

Reply:
> The point is that we don't really care if we can use userfaultfd unless user explicitly asked for lazy pages.

From other point of view, we might want to know early if kernel behaves strangely and after successful SYS_userfaultfd we don't expect to see uffd_api_ioctl, to fail and see bad uffdio api. We expect kernel to report absence of uffd gracefuly...

With the way it is done now, we would detect this strange behavior even if uffd is not actually used.